### PR TITLE
improve multi-core performance of qbatch_norm2d

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* #68328
* #72575
* #71903
* #69667
* #69601
* #69600
* __->__ #69599
* #69598
* #69517

Differential Revision: [D33353154](https://our.internmc.facebook.com/intern/diff/D33353154)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353154).

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 We have a few int8-related PRs targeting 1.11. And this is one of them. Could you prioritize to review these PRs? Thanks.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353154).

**Comment by mingfeima** [comment]
### Updates of this PR
improves `qbatch_norm` multi-core performance.

the original kernel will transform input to channels last memory format, and {N, (D), H, W} is collapsed as the outer dimension, {C} as the inner dimension. This PR parallels on outer dimension while the original kernel go sequentially.

### Benchmarking

* **short tag**: `python -m pt.qbatchnorm_test`
* **long tag**:  qbatchnorm_test doesn't have a long tag, pass.

CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, dual sockets, 20 cores per socket.
Unit: us

#### 1.a) single core run on short tag
this PR has no effect for single core run.

name | before | after | speedup
-- | -- | -- | --
QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8 | 1070.564 | 1062.101 | 100.80%
QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8 | 520.147 | 500.845 | 103.85%



#### 1.b) single socket run on short tag
the benchmark give inputs in contiguous memory format which will need to be transposed to channels last in the qbatch_norm kernel, this will eat up some of the perf speedup. In real model, batchnorm is already in channels last (if not already folded).


name | before | after | speedup
-- | -- | -- | --
QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8 | 288.736 | 150.749 | 191.53%
QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8 | 202.791 | 55.8 | 363.42%

**Comment by dzdang** [comment]
> ### Updates of this PR
> improves `qbatch_norm` multi-core performance.
> 
> the original kernel will transform input to channels last memory format, and {N, (D), H, W} is collapsed as the outer dimension, {C} as the inner dimension. This PR parallels on outer dimension while the original kernel go sequentially.
> 
> ### Benchmarking
> * **short tag**: `python -m pt.qbatchnorm_test`
> * **long tag**:  qbatchnorm_test doesn't have a long tag, pass.
> 
> CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, dual sockets, 20 cores per socket. Unit: us
> 
> #### 1.a) single core run on short tag
> this PR has no effect for single core run.
> 
> name	before	after	speedup
> QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8	1070.564	1062.101	100.80%
> QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8	520.147	500.845	103.85%
> #### 1.b) single socket run on short tag
> the benchmark give inputs in contiguous memory format which will need to be transposed to channels last in the qbatch_norm kernel, this will eat up some of the perf speedup. In real model, batchnorm is usually already in channels last (if not already folded).
> 
> name	before	after	speedup
> QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8	288.736	150.749	191.53%
> QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8	202.791	55.8	363.42%

The QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8	& QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8 speedups seem incorrect, or is it the After that is incorrect?

**Comment by mingfeima** [comment]
> ### Updates of this PR
> improves `qbatch_norm` multi-core performance.
> 
> the original kernel will transform input to channels last memory format, and {N, (D), H, W} is collapsed as the outer dimension, {C} as the inner dimension. This PR parallels on outer dimension while the original kernel go sequentially.
> 
> ### Benchmarking
> * **short tag**: `python -m pt.qbatchnorm_test`
> * **long tag**:  qbatchnorm_test doesn't have a long tag, pass.
> 
> CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, dual sockets, 20 cores per socket. Unit: us
> 
> #### 1.a) single core run on short tag
> this PR has no effect for single core run.
> 
> name	before	after	speedup
> QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8	1070.564	1062.101	100.80%
> QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8	520.147	500.845	103.85%
> #### 1.b) single socket run on short tag
> the benchmark give inputs in contiguous memory format which will need to be transposed to channels last in the qbatch_norm kernel, this will eat up some of the perf speedup. In real model, batchnorm is usually already in channels last (if not already folded).
> 
> name	before	after	speedup
> QBatchNorm1d_M1_N256_K3136_cpu_dtypetorch.qint8	288.736	150.749	191.53%
> QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8	202.791	55.8	363.42%

  @dzdang  Oh sorry, i missed the message... I think my notification mail box is messed up again...

The benchmark is comparing against of `QBatchNorm1d` of shape {M, N, K} and `QBatchNorm2d` of shape {M, N, K, 1}.
Theoretically the amount of computation is the same, but `qbatch_norm1d` actually takes more than `qbatch_norm2d`:
* `qbatch_norm1d`: a) 3d to 4d; b) make it nhwc contiguous; c) compute; d) make it nchw contiguous; e) 4d to 3d
* `qbatch_norm2d`: a)make it nhwc contiguous; b) compute

reference: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp#L147

I checked with VTune and actually memory format conversion takes a big part, so this benchmark is comparing "memory format conversion" + "qbatch_norm", something like this (the 1st row is triggered by memory format conversion and only the 3rd row is qbatch_norm kernel):

qbatch_norm1d_c1

And qbatch_norm1d has 2 conversion while qbatch_norm2d has 1. So theoretically qbatch_norm1d should have less speedup compared with qbatch_norm2d since memory format conversion takes more time in   qbatch_norm1d.

i hacked into the benchmark script to make 2d test in channels last (can't do this on qbatch_norm1d since channels last only applies to 4d tensors) and the result looks like:

```bash
### update performance with this patch on single core and multi core

### single core
# Benchmarking PyTorch: QBatchNorm2d
# Mode: Eager
# Name: QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8
# Input: M: 1, N: 256, K: 3136, device: cpu, dtype: torch.qint8
Forward Execution Time (us) : 132.453

### multi core
# Benchmarking PyTorch: QBatchNorm2d
# Mode: Eager
# Name: QBatchNorm2d_M1_N256_K3136_cpu_dtypetorch.qint8
# Input: M: 1, N: 256, K: 3136, device: cpu, dtype: torch.qint8
Forward Execution Time (us) : 22.004
```
There you see if only take the qbatch_norm kernel into account, will have even bigger speedup ratio.

**Review by dzdang — APPROVED** [review]

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353154).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33353154).
